### PR TITLE
[SPARK-48712][SQL] Perf Improvement for encode with empty values or UTF-8 charset

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR makes a short-circuit, which gets the underlying byte array directly and bypasses the encoding progress, for 'encoding UTF8String instances w/ UTF-8 charset' or 'UTF8String.EMPTY_STRING w/ any charset'.

### Why are the changes needed?

Performance improvement, 10x~20x according to benchmark results

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new unit tests
benchmark tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @HyukjinKwon @dongjoon-hyun please help review this PR when you are available, thank you in advance

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master
Thank you @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -3027,6 +3027,9 @@ object Encode {
       legacyCharsets: Boolean,
       legacyErrorAction: Boolean): Array[Byte] = {
     val toCharset = charset.toString
+    if (input.numBytes == 0 || "UTF-8".equalsIgnoreCase(toCharset)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3030]: this is actually a behavior change. If the input bytes are not valid utf 8 encoding, previously the result was not the same as the input bytes, but now it is.

We should either remove this utf 8 shortcut, or check the input bytes to see if it's valid utf8 encoding first.

cc @yaooqinn

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3030]: Do you mean that we will encode the unmappable characters to mojibakes before this PR, but now we use its identity?

Do you think we can call `input.isValid` to check here?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:3030]: yea I think so. For the happy path it's still faster than doing the actual encoding, and invalid utf8 bytes should be rare so it's ok to have an extra `isValid` call.

**Pull request closed without merging**